Improvements in the packet_device module

t0mk opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This PR improves the packet_device module.
- Bump of packet-python to 1.35.
- Changed module parameter `wait` to `wait_for_public_IPv`. It's not required, and it can be 4 or 6. When `wait_for_public_IPv: 4`, the module will wait until the new device gets public IPv4 address. Similarly for IPv6. When not set, the module will not wait for the automatic IP assignment.
- Improved handling of `state: active`. In the Packet Host API, a newly allocated device is in `provisioning` state for a few minutes, before it's `active`. Many API operation fail when the device is not yet `active`, so I made the module wait for the respective API state when `state: active` is passed to the module. I increased the timeout as this takes a few minutes.
- changed (fixed) argument 'lock' to 'locked'
- iPXE boot arguments.
- NB: with the default state `present`, the module returns right after the device-creating HTTP request succeeds.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
modules/cloud/packet/packet_device.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (packet-device-rework fc8c52afac) last updated 2017/03/30 15:04:06 (GMT +300)
  config file =
  configured module search path = [u'/home/tomk/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

t0mk committed:
Improvements and fixes in the packet_device module

t0mk commented:
Hi, could any of the maintainers please look at this?
I rebased this PR to newest devel, and did some fixes for pep8.
This is quite important fixup of the initial version of the packet_device module. There's been several users asking for these improvements on the Packet Slack channel - now we have to refer them to this PR, so I think it would be nice to correct the version in Ansible upstream.

I am open to maintainers comments and suggestions, and happy to fix those straight away.

t0mk commented:
ready_for_review

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -415,35 +437,52 @@ def create_single_device(module, packet_conn, hostname):
+        if all(d.state == 'active' for d in refreshed):
+            return refreshed
+        time.sleep(5)
+    raise Exception("Waiting for state \"active\" timed out for devices: %s"
```
It seems the exception message contains all watched devices instead of not `active` device, should not that be: ```% [d.hostname for d in refreshed if d.state != 'active'])``` (if so `refreshed` must be initialized before `while` block) ?

pilou- replied:
When `wait_timeout` is `0` the exception message is `Waiting for state \"active\" timed out for devices: []"`, what do you think of initializing `refreshed` with `watched_devices` instead of an empty list ?

abadger replied:
@t0mk @pilou- github doesn't always show me what's going on... were these comments addressed?

pilou- replied:
Yes, they were.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -415,35 +437,52 @@ def create_single_device(module, packet_conn, hostname):
+
+    def has_public_ip(addr_list, ip_v):
+        return any([a['public'] and a['address_family'] == ip_v and
+                    (len(a['address']) > 0) for a in addr_list])
```
`a['adress']` could be used instead of `(len(a['address']) > 0)`.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -562,10 +619,17 @@ def main():
 
+    except Exception as e:
+        module.fail_json(msg='failed to set machine state %s, error: %s'
+                         % (state, str(e)))
```
`str` is superfluous here, it should be removed.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -562,10 +619,17 @@ def main():
+        # e.cause is a requests.exceptions.HTTPError
+        msg = dig_error_json(e.cause)
+        if not msg:
+            msg = 'failed to set machine state %s, error: %s' % (state, str(e))
```
`str` is superfluous here, it should be removed.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -477,46 +516,64 @@ def act_on_devices(target_state, module, packet_conn):
+    # the errors from response JSON are much more helpful than just 422
+    resp = http_err.response
+    if resp is not None:
+        if resp.json():
```
In which case is `if resp.json():` useful ?

t0mk replied:
I improved the comment in the function.
The json is useful when the error comes from constraints of the Packet resources (e.g., trying to switch-off a device in "provisioning" state).
I check for the json it defensively, in order to handle also HTTP responses that don't contain the error descriptions.

pilou- replied:
I checked again, `json()` can return `None`, you're right, this check is useful. You might want to add a `try/except ValueError` around `resp.json()` (requests [doc](http://docs.python-requests.org/en/master/api/#requests.Response.json)).

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -368,7 +391,7 @@ def get_hostname_list(module):
 
     if len(hostnames) > MAX_DEVICES:
         raise Exception("You specified too many devices, max is %d" %
```
Should not it be `You specified too many hostnames` ?

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -100,17 +102,19 @@
     description:
-      - How long to wait for IP address of new devices before quitting. In seconds.
-    default: 60
+      - How long to wait either for automatic IP address assignment, or for the device to reach the C(active) state.
```
I guess this is `seconds` but unit must be added.

When `state` is `active` and `wait_for_public_IPv` is set, then value of `wait_timeout` is used twice (hence max waiting time is 2*`wait_timeout`): it should be specified.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -91,6 +91,8 @@
     description:
       - Desired state of the device.
+      - If set to C(present) (the default), the module call will return immediately after the device-creating HTTP request successfully returns.
+      - If set to C(active), the module call will block until all the specified devices are in state active due to the Packet API, which might take few minutes.
```
When `state` is `active`, `wait_timeout` is applied and the call won't block more than `wait_timeout`. The documentation of `state`must be updated too.

pilou- reviewed (commented):
Have you considered adding unit or integration tests for this module ?

pilou- commented:
@t0mk Here is a list of minor issues unrelated to your commits. I mention them because this pull request already contains some cleaning and `DOCUMENTATION` fixes.

* in `listify_string_name_or_id`, method use `s.split(',')` instead of `[i for i in s.split(',')]`
* there is a problem with [`lock`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L426)/[`locked`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L428) parameter. DOCUMENTATION uses [`lock`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L73), [`EXAMPLES`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L176) uses `locked`, code use both.
* `DOCUMENTATION`
  * `requirements` field, remove `"python >= 2.6"` (see [roadmap](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_4.rst#python-2-4-and-2-5-support-discontinuation))
  * add default values for `count` (`1`), `count_offset` (`1`) and `facility` (`ewr1)
  * `hostnames` description: use `I(count)` instead of `count` ([see](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options))
  * `features` / `plan` / `operating_system`: if available, add a link to Packets documentation (using formatting function)
  * add in `notes` field that `check_mode` isn't supported
  * `RETURN`: `devices` isn't returned in case of failure (`devices` isn't always returned, `returned` must be updated)
  * you could remove default settings (`required: false`, `default: None`)

t0mk commented:
@pilou- Thanks for your review! I agree with most things, and I fixed them.

Integration test is a good idea, I wanted to do it myself before, but found the testing docs a bit chaotic. I will try to put one together.

t0mk commented:
I added an integration test in 23b6ecd. @pilou- I noticed you are member of the testing group, can you give me feedback on the test?

In order to run it, one needs to add packet_api_token and packet_test_project_id to integration_config.yml, and then run

```
$ ansible-test --local -v packet_device
```

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -91,30 +94,35 @@
-      - How long to wait for IP address of new devices before quitting. In seconds.
-    default: 60
+      - How long (secods) to wait either for automatic IP address assignment, or for the device to reach the C(active) I(state).
+      - If I(wait_for_public_IPv) is set and I(state) is C(active), the module will wait for both events consequently, applying the timout twice.
```
s/timout/timeout/

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -91,30 +94,35 @@
     description:
-      - How long to wait for IP address of new devices before quitting. In seconds.
-    default: 60
+      - How long (secods) to wait either for automatic IP address assignment, or for the device to reach the C(active) I(state).
```
s/secods/seconds/

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -91,30 +94,35 @@
+    default: 900
 
 requirements:
      - packet-python
```
You could add the minimum required version (`>= 1.0`) here too.

pilou- reviewed (commented):
Great ! I will review the test part before end of the week.

t0mk commented:
@pilou- The test that I wrote needs valid API credentials and, at least at this point, can't be run from Ansible CI at shippable.

Should I move it to `test/integration/roles/test_packet_device` and then write a testing playbook `test/integration/packet.yml`, and write entry for the playbook to `test/integration/Makefile`?

pilou- commented:
> The test that I wrote needs valid API credentials and, at least at this point, can't be run from Ansible CI at shippable.

Integration tests could be modified in order to only check that authentication fails (something like [that](https://github.com/t0mk/ansible/blob/438c2079c4c3912b19fb221a8a6f2000204352cf/test/integration/targets/setup_ec2/tasks/common.yml)). In addition, unit tests (using mock) could be used.

> Should I move it to `test/integration/roles/test_packet_device` and then write a testing playbook `test/integration/packet.yml`, and write entry for the playbook to `test/integration/Makefile`?

As far as I known, no. Playbooks in `test/integration/*.yml` are legacy (they existed before ansible-test) or special cases for network testing. New integration tests go into `test/integration/targets`. As the bot wrote, a `test/integration/targets/packet_device` should be added, it seems this file should contains something like `cloud/packet`.

t0mk commented:
@pilou- That's a pretty bad news - test for auth fail does not really test functionality of the packet_device module. And IMO unit test for a cloud module has very little sense - it will never properly test the real and relevant interaction with the API. Indeed `test/units/modules/cloud` is quite sparse.

Maybe I could ask Packet people for credentials dedicated to Ansible integration tests. Are the integration tests for a particular cloud module run in the CI on every single commit, or just commits changing the particular module?

I can't imagine Packet host would be comfortable having their API hit on every single Ansible commit.

With the current situation, I think it's best if I just remove the test from this PR. It's pity because an integration test playbook (run on demand with valid credentials) would be a really good resource also for learning advanced usage of the module.

pilou- commented:
> Maybe I could ask Packet people for credentials dedicated to Ansible integration tests. Are the integration tests for a particular cloud module run in the CI on every single commit, or just commits changing the particular module?

Ansible Packet tests would be run when a pull request updating the Ansible Packet module would be created or updated and other pull requests impacting all modules would be created or updated. Ansible Packet tests should not be executed when pull requests related to other modules are created or executed.

gundalow commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -22,19 +22,19 @@
+    - This module has a dependency on packet >= 1.0.
+    - API is documented at U(https://www.packet.net/developers/api/devices).
 
-version_added: 2.3
```
I don't believe this should have changed. The module was available in 2.3

gundalow commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -66,22 +68,23 @@
     description:
       - Whether to lock a created device.
     default: false
+    version_added: 2.4
```
You can remove this line, you will get a CI failure, though that will go away once this PR is merged

t0mk commented:
@gundalow , just fixed your latest comments.

I also removed the integration test. I will contribute it as separate PR, once we, hopefully, add credentials for Packet integration tests to Ansible CI. I am actually very curious on how to proceed with that & we have green light on that from Packet people - they're ready to provide dedicated account for Ansible integration test. 

You or @pilou-: please let me know if there's anything you'd need from my side.

pilou- commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -477,46 +520,66 @@ def act_on_devices(target_state, module, packet_conn):
     }
 
 
+def dig_error_json(http_err):
```
Oh, this method seems a workaround of a bug in the Packet library. Indeed the library [already search](https://github.com/packethost/packet-python/blob/master/packet/baseapi.py#L83) for `error` key in response and use it to raise an exception but in fact this line is never reached because `raise_for_status` is [called before](https://github.com/packethost/packet-python/blob/master/packet/baseapi.py#L67). `ok` property [returns `False`](https://github.com/requests/requests/blob/master/requests/models.py#L687) only when `raise_for_status` raise an exception.

t0mk replied:
Yes the error handling there is not so pretty. Maybe I will try to fix it and submit PR to packet-python. This workaround seemed less complicated than going through the exception flow in packet-python, fixing and waiting.

t0mk replied:
I fixed it in https://github.com/packethost/packet-python/pull/26

pilou- replied:
It is fixed and released, should not `dig_error_json` be removed (or called only when `packet-python` version is older than 1.34) ?

t0mk commented:
@gundalow @pilou- I got input from last Ansible testing group meeting and the integration test problem for Packet resources seems over the scope of this PR.

Can we refocus on reviewing the fixes on the packet_device module introduced by this PR? I think it would be useful for users to have this in, and it's blocking other improvements we have planned.

If you agree with me that integration test is not a possibility here, what else I can do to get this merged? I am against unit tests, since I think they don't really add any value in cases like this, when the API is subject to change, volatile, and out of control of the test author.

At the moment, the CI fails because I removed the version_added items for new module parameters, according to @gundalow's review.

gundalow commented:
+1 to getting this PR in *before* the 2.4 feature freeze. We can look at testing afterwards

t0mk commented:
I squashed the commits.

pilou- commented:
About tests: this module doesn't call directly the HTTP API (for example it's not possible to set endpoint), the module relies on [`packet-python`](https://pypi.python.org/pypi/packet-python) in order to call Packet API.
Hence:
1. this module could trust and rely on tests belonging to `packet-python`; if the webservice evolves with backward incompatible changes, then `packet-python` would need to be updated (but not necessarily this Ansible module).
2. unit tests make sense for this Ansible module

@t0mk you should comment this pull request with [`ready_for_review`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#cmd-ready_for_review) comment in order to get rid of [`needs_revision`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#label-needs_revision) label. But there isn't any maintainer for this module (neither module maintainer nor namespace maintainer), a way could be to add this pull request to the [public core meeting agenda](https://github.com/ansible/community/issues/220). [Meetings](https://github.com/ansible/community/blob/master/meetings/core-team.yaml) are Tuesday 19:00 UTC and Thursday 15:00 UTC.

t0mk commented:
I fixed the error handling (after my PR fixing the root cause in packet-python was accepted, and packet-python re-released), I also fixed it in all the other pending PRs for packet modules.

I am not going to write any unit test for cloud modules, because I think it's actually counterproductive. I wrote the reasons above, and in email. I do want to test the modules, but in a useful way - I put integration tests for packet modules to this repo: https://github.com/t0mk/packet-ansible-tests. I will use those for checking the stability of module after changes.

t0mk commented:
ready_for_review

bcoca commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -22,17 +22,17 @@
-    - create, destroy, update, start, stop, and reboot a Packet Host machine. When the machine is created it can optionally wait for it to have an
-      IP address before returning. This module has a dependency on packet >= 1.0.
-    - API is documented at U(https://www.packet.net/help/api/#page:devices,header:devices-devices-post).
+    - create, destroy, update, start, stop, and reboot a bare metal server in the Packet Host (a "device" in the API terms).
```
'manage' should cover all actions, no need to be this explicit in the description

abadger commented on lib/ansible/modules/cloud/packet/packet_device.py:
```diff
@@ -562,10 +614,10 @@ def main():
-        module.fail_json(msg='failed to set machine state %s, error: %s' % (state,str(e)))
-
+        module.fail_json(msg='failed to set machine state %s, error: %s'
+                         % (state, e))
```
When catching a generic exception, not knowing why it failed; you should include the stacktrace.  It's also safer to manually transform the exception into a strig usig to_native.  So this code will look like this:
``` python
import traceback

from ansible.module_utils._text import to_native
[...]

module.fail_json(msg='failed to set machine state %s, error: %s'
                           % (state, to_native(e)), exception=traceback.format_exc())

abadger commented:
These changes break backwards compatibility.  The module is status='preview' but we've not decided how to decide if compatibility breaks are allowed for that status.  This was brought up at the IRC meeting today but no decision was reached.

* Everyone agreed that backwards incompatibilities need to be documented.  So that's a concrete change to this PR that can be made.
* Several proposals were made on how to decide:
  * Leave it to the module maintainers (so the authors of the packet_device module) to discuss and decide
  * Leave it to the domain experts (so the authors of modules that live inside of cloud/packet) to discuss and decide
  * Decide on something(?) project-wide with case-by-case exceptions

I'll put this question as those proposals onto the agenda for Tuesday.

t0mk commented:
@abadger @bcoca Thanks for your review comments, I will fix them tomorrow morning european time.

Ad breakign backward compatibility: I am happy to document it. Would you have any example on how t document it? Link to a module doc or a guide?

abadger commented:
@t0mk Here's one from the unarchive module https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/unarchive.py#L41

t0mk commented:
@abadger, I fixed your comments, and documented the deprecated arguments.

I also added two options for iPXE boot of Packet devices, it's a small change.

t0mk commented:
@abadger, I fixed your review comments. The exception handling cause the merge conflict.

I added options for iPXE boot for device in Packet host, it's a small change. Then I squashed the commits.

jctanner commented:
@t0mk this needs a rebase

t0mk commented:
@jctanner, I rebased.

abadger commented:
@t0mk The above errors from shippable mean that you need to add version_added: 2.4 to the documentation for those options.

t0mk committed:
add version_added to new args

t0mk commented:
@abadger I know - I cleared the version_added there because @gundalow wrote that I can remove it. I guess it was because he wasn't sure if this will go to 2.4.:
https://github.com/ansible/ansible/pull/23127#discussion_r131205111

Anyway, I re-added the version_added to the new args. I'd be grateful for concise and consistent instructions on improving this PR.

abadger commented:
Code-wise this looks ready.  We'll get back to you on whether we want something more for backwards incompatibility no later than Tuesday next week (our next public IRC meeting)

t0mk commented:
Thanks!

On Sun, Aug 13, 2017 at 6:21 PM, Toshio Kuratomi <notifications@github.com>
wrote:

> Code-wise this looks ready. We'll get back to you on whether we want
> something more for backwards incompatibility no later than Tuesday next
> week (our next public IRC meeting)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/23127#issuecomment-322048240>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKr_aiFFiyDYGDM20XnKwJbTEmhpAlsks5sXxRegaJpZM4MuSWY>
> .
>

abadger commented:
@t0mk I noticed one set of comments from @pilou- that don't seem to be addressed.  can you double check that for me?  Otherwise this is good to go.

Results of today's discussion on backwards compat: tldr; Nothing further is needed

We decided that when and how to break backwards compat is largely up to the maintainers with jsut two rules: (1) incompatibilities need to be doumented (which has been done here) and (2) they can't be done in a minor release.  We're getting ready to ship 2.4 from devel so merging to devel now is also fine.

pilou- commented:
Two notes:
- default value (`ewr1`) for `facility` parameter is not documented
- `lock` parameter has been removed in the code but the documentation states that this parameter has been deprecated. It seems deprecated means _it's available but don't use it_. Should not `lock` parameter documentation removed? `lock` parameter documentation could be removed and `lock` could be added as an alias of `locked`.

t0mk committed:
remove default value from facility

t0mk commented:
I documented the `lock` as deprecated param, just like it was suggested by @abadger  at https://github.com/ansible/ansible/pull/23127#issuecomment-321713397 and analogically to the `copy` deprecated option here:
https://github.com/ansible/ansible/blob/ce3d1c6ba00e1a36c016575ea7d7d04b71923486/lib/ansible/modules/files/unarchive.py#L41

I removed the defualt value of facility, it's better that way.

pilou- commented:
In Ansible `core`, deprecated means `available but will be removed` (for example: http://docs.ansible.com/ansible/latest/porting_guide_2.0.html#deprecated), I guess users could be surprised.

~~Anyway, let's move forward :)~~
~~ship it~~

Edited: @t0mk `copy` is still a usable parameter of `unarchive` (see 4ac135c1b5ed6a7b9e3151b33c1f38a85f527465). Hence `lock` parameter thing need to be fixed.

t0mk committed:
changed 'lock' from deprecated arg to alias of 'locked'

t0mk commented:
@pilou- I changed the "lock" param to alias of "locked".

pilou- commented:
Thanks !

shipit

@gundalow, @abadger could you merge this one ?

abadger commented:
Merged to devel for 2.4.0
